base16ct/src/lib.rs

```diff
@@ -0,0 +1,228 @@
+    pub fn lower_case() -> Base16 {
+        Base16 { case: Case::Lower }
+    }
+
+    /// Upper case hex: 0-9 A-F
+    pub fn upper_case() -> Base16 {
+        Base16 { case: Case::Upper }
+    }
+}
+
+impl Base16 {
+    /// Decode a Base16 (hex) string into the provided destination buffer.
+    pub fn decode<'a>(&self, src: impl AsRef<[u8]>, dst: &'a mut [u8]) -> Result<&'a [u8], Error> {
+        let src = src.as_ref();
+
+        let dst_length = Self::decoded_len(src)?;
+        if dst_length > dst.len() {
+            return Err(Error::InvalidLength);
+        }
+
+        let mut err: usize = 0;
+
+        for (i, dst_byte) in dst.iter_mut().enumerate().take(dst_length) {
+            let src_offset = i * 2;
+            let byte = (self.case.decode_nibble(src[src_offset]) << 4)
+                | self.case.decode_nibble(src[src_offset + 1]);
+            err |= byte >> 8;
+            *dst_byte = byte as u8;
+        }
+
+        if err == 0 {
+            Ok(&dst[..dst_length])
+        } else {
+            Err(Error::InvalidEncoding)
+        }
+    }
+
+    /// Decode a Base16 (hex) string into a byte vector.
+    #[cfg(feature = "alloc")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+    pub fn decode_vec(&self, input: impl AsRef<[u8]>) -> Result<Vec<u8>, Error> {
+        let mut output = vec![0u8; Self::decoded_len(input.as_ref())?];
+        self.decode(input, &mut output)?;
+        Ok(output)
+    }
+
+    /// Compute decoded length of the given input.
+    pub fn decoded_len(bytes: &[u8]) -> Result<usize, Error> {
+        if bytes.len() & 1 == 0 {
+            Ok(bytes.len() >> 1)
+        } else {
+            Err(Error::InvalidLength)
+        }
+    }
+
+    /// Encode the input byte slice as Base16 (hex).
+    ///
+    /// Writes the result into the provided destination slice, returning an
+    /// ASCII-encoded Base16 (hex) string value.
+    pub fn encode<'a>(&self, src: &[u8], dst: &'a mut [u8]) -> Result<&'a [u8], Error> {
+        if Self::encoded_len(src) > dst.len() {
+            return Err(Error::InvalidLength);
+        }
+
+        for (i, src_byte) in src.iter().enumerate() {
+            let offset = i * 2;
+            dst[offset] = self.case.encode_nibble(src_byte >> 4);
+            dst[offset + 1] = self.case.encode_nibble(src_byte & 0x0f);
+        }
+
+        Ok(&dst[..(src.len() * 2)])
+    }
+
+    /// Encode input byte slice into a [`String`] containing Base16 (hex).
+    ///
+    /// # Panics
+    /// If `input` length is greater than `usize::MAX/2`.
+    #[cfg(feature = "alloc")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+    pub fn encode_string(&self, input: &[u8]) -> String {
+        let elen = Self::encoded_len(input);
+        let mut dst = vec![0u8; elen];
+        let res = self.encode(input, &mut dst).expect("encoding error");
+
+        debug_assert_eq!(elen, res.len());
+        String::from_utf8(dst).expect("character encoding error")
```

Comment:
I think we can use `unsafe { String::from_utf8_unchecked(dst) }` like we do in `base64ct` to remove unreachable panic.